File: src/bot.py
from selenium import webdriver
from selenium.webdriver.common.keys import Keys
from getpass import getpass
from saveData import BotData
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions
import time

class Bot:
	PATH = './src/chromedriver.exe'
	url = 'https://instagram.com'
	custom_url = 'https://www.instagram.com/{username}/'
	data: BotData
	totalFound = 0

	def __init__(self, username: str, password: str, data: BotData):
		print('initialize bot')
		self.username = username
		self.password = password
		self.data  = data
		self.exec_time = time.time()

	def open(self):
		mobile_emulation = { "deviceName": "Nexus 5" }
		chrome_options = webdriver.ChromeOptions()
		chrome_options.add_experimental_option("mobileEmulation", mobile_emulation)
		self.driver = webdriver.Chrome(self.PATH, options=chrome_options)
		self.driver.get(self.url)
		time.sleep(1)

	def __handleNotificationRequest(self):
		try:
			not_now_btn = driver.find_element_by_xpath('/html/body/div[4]/div/div/div[3]/button[2]')
			not_now_btn.click()
			print('not now')
		except:
			pass

	def __clickLoginButton(self):
		try:
			self.driver.find_element_by_xpath("//body/div[2]/div[1]/div[1]/div[1]/div[2]/button[1]").click()
		except:
			pass

		self.driver.find_element_by_xpath("//*[@id='react-root']/section[1]/main[1]/article[1]/div[1]/div[1]/div[1]/div[2]/button[1]").click()

	def doLogin(self):
		self.__clickLoginButton()
		USERNAME = self.driver.find_element_by_name('username').send_keys(self.username)
		PASSWORD = self.driver.find_element_by_name('password').send_keys(self.password)

		login_btn = self.driver.find_element_by_xpath("//*[@id='loginForm']/div[1]/div[6]/button[1]").click()
		time.sleep(5)
		self.driver.get(self.url)
		self.__handleNotificationRequest()

	def goToDirect(self):
		self.driver.get("https://www.instagram.com/direct/inbox")
		self.driver.find_element_by_xpath("//body/div[4]/div/div/div/div[3]/button[2]").click()
		self.driver.find_element_by_xpath("//body").click()
		self.driver.find_element_by_xpath("//*[@id='react-root']/section/div[1]/header/div[1]/div[2]/button").click()

	def sendMessage(self, username: str, message: str):
		lclock = time.time()
		search_box = self.driver.find_element_by_xpath("//*[@id='react-root']/section[1]/div[2]/div[1]/div[1]/div[1]/div[2]/input[1]")
		search_box.click()
		search_box.send_keys(username)
		time.sleep(2.5)
		# click profile
		self.driver.find_element_by_xpath("//*[@id='react-root']/section[1]/div[2]/div[1]/div[2]/div[1]/div[1]/div[3]/button[1]").click()
		
		# click "avanti"
		WebBrowserWait(self.driver, 5).until(EC.element_to_be_clickable(By.XPATH, "//*[@id='react-root']/section[1]/div[1]/header/div/div[2]/button"))
		self.driver.find_element_by_xpath("//*[@id='react-root']/section[1]/div[1]/header/div/div[2]/button").click()

		self.exec_time = (time.time() - self.start_time)/60
		if self.exec_time % 20 == 0: time.sleep(60)
		if self.exec_time % 120 == 0: time.sleep(60 * 17)
		else: time.sleep(random.uniform(4.0, 7.5))
		# click textarea
		self.driver.find_element_by_xpath("//*[@id='react-root']/section[1]/div[2]/div[1]/div[1]/div[2]/div[1]/div[1]/div[1]/textarea[1]").click()

		if message != "":
			self.driver.find_element_by_xpath("//*[@id='react-root']/section[1]/div[2]/div[1]/div[1]/div[2]/div[1]/div[1]/div[1]/textarea[1]").send_keys(message)
			self.driver.find_element_by_xpath("//*[@id='react-root']/section[1]/div[2]/div[1]/div[1]/div[2]/div[1]/div[1]/div[2]/button[1]").click()
		self.driver.find_element_by_xpath("//*[@id='react-root']/section[1]/div[1]/header[1]/div[1]/div[1]").click()

		self.data.makeDone(username)
		time.sleep(0.5)
		print('This user took about: ' + (time.time() - lclock))

	def scrollDown(self):
		last_height = self.driver.execute_script("return document.body.scrollHeight")

		while True:
				# Scroll down to bottom
				self.driver.execute_script("window.scrollTo(0, document.body.scrollHeight);")

				# Wait to load page
				time.sleep(2)

				# Calculate new scroll height and compare with last scroll height
				new_height = self.driver.execute_script("return document.body.scrollHeight")
				if new_height == last_height:
					break
				last_height = new_height

	def searchHashtag(self, hashtag: str):
		custom_url = self.url + '/explore/tags/{tag}'
		custom_url = custom_url.format(tag = hashtag)
		self.driver.get(custom_url)

	def searchUser(self, username: str):
		custom_url = self.custom_url.format(username = username)
		self.driver.get(custom_url)

	def addFollowedToList(self):
		# click "seguiti"
		self.driver.find_element_by_xpath("//*[@id='react-root']/section[1]/main[1]/div[1]/ul[1]/li[3]/a[1]/span[1]").click()
		# add account to list

		self.scrollDown()

		WebDriverWait(self.driver, 10).until(expected_conditions.visibility_of_element_located((By.CSS_SELECTOR, 'a.notranslate')))
		list_of_elements = self.driver.find_elements_by_css_selector('a.notranslate')
		print(list_of_elements)
		for i in list_of_elements:
			self.totalFound += 1
			print (i.text)
			self.data.addFound(i.text)

		print('found {count} new accounts'.format(count = self.totalFound))

	def addFollowersToList(self):
		# click "seguiti"
		self.driver.find_element_by_xpath("//*[@id='react-root']/section[1]/main[1]/div[1]/ul[1]/li[2]/a[1]/span[1]").click()
		# add account to list

		self.scrollDown()

		WebDriverWait(self.driver, 10).until(expected_conditions.visibility_of_element_located((By.CSS_SELECTOR, 'a.notranslate')))
		list_of_elements = self.driver.find_elements_by_css_selector('a.notranslate')
		print(list_of_elements)
		for i in list_of_elements:
			self.totalFound += 1
			print (i.text)
			self.data.addFound(i.text)

		print('found {count} new accounts'.format(count = self.totalFound))


	def close(self):
		self.driver.close()

File: src/bot.ts
import * as puppeteer from 'puppeteer';


class Bot {
  chromedriver_path: string = 'C:/dev/instapy/bin/';
  intagram_url: string = 'https://instagram.com/';
  username: string | null = null;
  password: string | null = null;
  iPhone = puppeteer.devices['iPhone 6'];
  driver = {};
  page: puppeteer.Page | null = null;
  start_time: number = Date.now();

  constructor(username: string, password: string) {
    this.username = username;
    this.password = password;
  }

  async open(): Promise<void> {
    const browser = await puppeteer.launch({ headless: false });
    this.page = await browser.newPage();
    await this.page.emulate(this.iPhone);
    await this.page.goto(this.intagram_url);
    console.log(await this.page.title());
  }

  private async acceptCookies(): Promise<void> {
    var [accept] = await this.page.$x('//body/div[2]/div[1]/div[1]/div[1]/div[2]/button[1]')
    if (accept) {
      await accept.tap()
    }
    else console.error('Failed to click accept cookie button')
  }

  private async clickAccedi(): Promise<void> {
    var accedi = await this.page.waitForXPath('//*[@id=\'react-root\']/section[1]/main[1]/article[1]/div[1]/div[1]/div[1]/div[2]/button[1]', {
      timeout: 4000,
      visible: true
    });
    if(accedi) return await accedi.tap();
    else console.error('Unable to click accedi button')
  }

  private async handleNotificationRequest(): Promise<void> {
		var [not_now_button] = await this.page.$x('//body/div[4]/div/div/div/div[3]/button[2]')
    if (not_now_button) {
      await not_now_button.tap()
    }
    else console.error('Failed to click not now notification button')
  }

  private async saveAccessInfo(): Promise<void> {
		var [not_now_button] = await this.page.$x('/html/body/div[1]/section/main/div[1]/div[1]/div[1]/button[1]')
    if (not_now_button) {
      await not_now_button.tap()
    }
    else console.error('Failed to click not now save info button')
  }

  private async clickBody(): Promise<void> {
		var [body] = await this.page.$x('//*[@id=\'react-root\']/section[1]/div[2]/div[1]/div[1]')
    if (body) {
      await body.tap()
    }
    else console.error('Failed to click body')
  }

  async doLogin(): Promise<void> {
    await this.acceptCookies();
    await this.clickAccedi();
    await this.page.waitForSelector("[name='username']");
    await this.page.type("[name='username']", this.username);
    await this.page.keyboard.down('Tab');
    await this.page.keyboard.type(this.password);
    var [login] = await this.page.$x('//*[@id=\'loginForm\']/div[1]/div[6]/button[1]')
    if (login) await login.tap();
    else console.error('Failed to click login button')
    await this.page.waitFor(8000);
    await this.saveAccessInfo();
    await this.page.waitForNavigation({
      waitUntil: 'networkidle0',
    });
    
  }
  
  async goToDirect(): Promise<void> {
    await this.page.goto('https://www.instagram.com/direct/inbox');
    await this.handleNotificationRequest();
    await this.clickBody();
  }
  
  async sendMessage(username: string, message: string ): Promise<void> {
    const lclock: number = Date.now();
    // type on search box
    await this.page.waitForXPath('//*[@id=\'react-root\']/section[1]/div[2]/div[1]/div[1]/div[1]/div[2]/input[1]');
    await this.page.type('//*[@id=\'react-root\']/section[1]/div[2]/div[1]/div[1]/div[1]/div[2]/input[1]', username);
    // click first account
    var firsst_acc = await this.page.waitForXPath('//*[@id=\'react-root\']/section[1]/div[2]/div[1]/div[2]/div[1]/div[1]/div[3]/button[1]');
    if (firsst_acc) await firsst_acc.tap();
    else console.error('Failed to click first account or account does not exist');
    // click avanti
    var avanti = await this.page.waitForXPath('//*[@id=\'react-root\']/section[1]/div[1]/header/div/div[2]/button')
    if(avanti) await avanti.tap();
    else console.error('Failed to click avanti button');

    var exec_time = (Date.now() - this.start_time) / 60;
    if (exec_time % 20 == 0) await setTimeout(() => true, 60 * 1000);
    if (exec_time % 117 == 0) await setTimeout(() => true, 17 * 60 * 1000);
    else await setTimeout(() => true, (Math.floor(Math.random() * (7.5 - 4 + 1)) + 4) * 1000);

    // send mesage
    await this.page.waitForXPath('//*[@id= \'react-root\']/section[1]/div[2]/div[1]/div[1]/div[2]/div[1]/div[1]/div[1]/textarea[1]');
    await this.page.type('//*[@id=\'react-root\']/section[1]/div[2]/div[1]/div[1]/div[2]/div[1]/div[1]/div[1]/textarea[1]', message);
    var send = await this.page.waitForXPath('//*[@id=\'react-root\']/section[1]/div[1]/header[1]/div[1]/div[1]');
    if (send) await send.tap();
    else console.error('Failed to send message');

    // todo data makedone 

    await setTimeout(() => true, 2000);

    console.log(`This user took about ${Date.now() - this.start_time}`);
  }

  async scrollDown(): Promise<void> {
    console.log('scrolldown');
    var last_height = await this.page.evaluate(() => {
      return document.body.scrollHeight;
    });

    while (true) {
      var time = Date.now();
      while (true) {
        if (last_height < await this.page.evaluate(() => {return document.body.scrollHeight;})) {
          last_height = await this.page.evaluate(() => {return document.body.scrollHeight;});
          break;
        }
        if ((Date.now() - time) > 20 * 1000) return;
      }

      await this.page.evaluate(() => { window.scrollTo(0, document.body.scrollHeight)});
      await setTimeout(() => true, (Math.floor(Math.random() * (1.65 - 0.58 + 1)) + 0.58) * 1000);
      console.log('scrolling');
    }
  }

  async searchHashtag(hashtag: string): Promise<void> {
    const hashtag_url = this.intagram_url + 'explore/tags/' + hashtag;
    await this.page.goto(hashtag_url);
  }

  async searchUser(username: string): Promise<void> {
    const username_url = this.intagram_url + username;
    await this.page.goto(username_url)
  }

  async getFollowedNumber(): Promise<number> { 
    var followed = await this.page.waitForXPath('//*[@id=\'react-root\']/section[1]/main[1]/div[1]/ul[1]/li[3]/a[1]/span[1]');
    if(followed) return parseInt(await this.page.evaluate(el => el.textContent, followed))
    else console.error('Unable to get followed number')
  }

  async getFollowersNumber(): Promise<number> { 
    var followed = await this.page.waitForXPath('//*[@id=\'react-root\']/section[1]/main[1]/div[1]/ul[1]/li[2]/a[1]/span[1]');
    if(followed) return parseInt(await this.page.evaluate(el => el.textContent, followed))
    else console.error('Unable to get followers number')
  }

  async getFollowersList(): Promise<Array<string>> {
    var count: number = 0;
    var str_list: Array<string>;
    var followers_button = await this.page.waitForXPath('//*[@id=\'react-root\']/section[1]/main[1]/div[1]/ul[1]/li[2]/a[1]/span[1]');
    if(followers_button) await followers_button.tap();
    else console.error('Unable to click followers button')

    await setTimeout(async () => {
      await this.scrollDown();
    }, 6 * 1000);

    var list = await this.page.$$('a .notranslate');
    for(var i of list ){
      count ++;
      const username = await this.page.evaluate(el => el.textContent, i);
      str_list.push(username)
      console.log('Username: ' + username);
    }

    console.log('Found ' + count.toString() + 'accounts');
    return str_list;
  }

  async getFollowedList(): Promise<Array<string>> {
    var count: number = 0;
    var str_list: {list: string[], length: 0};
    var followers_button = await this.page.waitForXPath('//*[@id=\'react-root\']/section[1]/main[1]/div[1]/ul[1]/li[3]/a[1]/span[1]');
    if(followers_button) await followers_button.tap();
    else console.error('Unable to click followed button')
    console.log('clicked followed button');

    console.log('strarting timeout of 6 sec');
    await setTimeout(() => {}, 6 * 1000);
    console.log('finished timeout');
    console.log('scrolling..');
    await this.scrollDown();
    console.log('done');

    var list = await this.page.$$('li');
    console.log(list)
    var obj_list = { list: list, length: list.length}
    console.log(obj_list.list.length);


    for(var i = 0; i < obj_list.length; i++ ){
      var name = await obj_list.list[i].$eval('a.FPmhX', node => node.textContent);
      count ++;
      console.log('here');
      str_list.list.push(name)
      console.log(str_list);
      console.log('Username: ' + name);
    }

    console.log('Found ' + count.toString() + ' accounts');
    return str_list.list;
  }

  async close(): Promise<void> {
    await this.page.close();
  }
}


export default Bot